From 88368b41bfc50b6ef650685bab54b8bf63c1fefd Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Morten=20S=C3=B8rvig?= <morten.sorvig@qt.io>
Date: Fri, 19 Sep 2025 13:00:41 +0200
Subject: [PATCH 192/553] wasm: set max memory to 16GB for wasm64

User code can override with QT_WASM_MAXIMUM_MEMORY, but
we should set the default maximum, similar to what we do
for wasm32.

16GB is the largest value which Emscripten accepts for
WASM_MAXIMUM_MEMORY for wasm64.

Change-Id: I1e7288b19776233d800aa82a728b0041040c3fed
Reviewed-by: Even Oscar Andersen <even.oscar.andersen@qt.io>
Reviewed-by: Lorn Potter <lorn.potter@qt.io>
(cherry picked from commit 2f91d024a70698853b8cdb9658e5e11a6e1c385a)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 mkspecs/features/wasm/wasm.prf  | 6 +++++-
 src/corelib/Qt6WasmMacros.cmake | 6 +++++-
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/mkspecs/features/wasm/wasm.prf b/mkspecs/features/wasm/wasm.prf
index 4bbba8c2aa6..d3f2d5e761c 100644
--- a/mkspecs/features/wasm/wasm.prf
+++ b/mkspecs/features/wasm/wasm.prf
@@ -50,7 +50,11 @@ exists($$QMAKE_QT_CONFIG) {
     }
     EMCC_LFLAGS += -s INITIAL_MEMORY=$$INITIAL_MEMORY
     isEmpty(QT_WASM_MAXIMUM_MEMORY) {
-        MAXIMUM_MEMORY = 4GB # 32-bit max
+        contains(QMAKE_CFLAGS, -s MEMORY64=1) {
+            MAXIMUM_MEMORY = 16GB
+        } else {
+            MAXIMUM_MEMORY = 4GB
+        }
     } else {
         MAXIMUM_MEMORY = $$QT_WASM_MAXIMUM_MEMORY
     }
diff --git a/src/corelib/Qt6WasmMacros.cmake b/src/corelib/Qt6WasmMacros.cmake
index 032a28f0473..75a93f155e5 100644
--- a/src/corelib/Qt6WasmMacros.cmake
+++ b/src/corelib/Qt6WasmMacros.cmake
@@ -130,7 +130,11 @@ function(_qt_internal_wasm_add_target_helpers target)
         if(_tmp_maximumMemory)
             set(QT_WASM_MAXIMUM_MEMORY "${_tmp_maximumMemory}")
         elseif(NOT DEFINED QT_WASM_MAXIMUM_MEMORY)
-            set(QT_WASM_MAXIMUM_MEMORY "4GB")
+            if(WASM64)
+                set(QT_WASM_MAXIMUM_MEMORY "16GB")
+            else()
+                set(QT_WASM_MAXIMUM_MEMORY "4GB")
+            endif()
         endif()
         target_link_options("${target}" PRIVATE "SHELL:-s MAXIMUM_MEMORY=${QT_WASM_MAXIMUM_MEMORY}")
 
-- 
2.51.2

